### DISCUSSION TITLE
lifetime of msghdr in sendmsg/sendmsg_zc

### DISCUSSION THREAD

**Discussion by pyhd** [axboe/liburing#1259] [Q&A] [answered]
As I read the code, each `msghdr` will be copied to `io_async_msghdr` after `io_uring_submit` except `-ENOMEM`. In other words, the userspace can reuse them immediately after a successful submission. Then I just wonder if there are any other exceptions.

**Comment by isilence** [comment, marked as answer]
It's not an exception, sendmsg[_zc] never accesses user's msghdr and iovec after the request has been submitted, i.e. `io_uring_submit` returns, regardless of the result.

**Reply by pyhd** [reply]
Thank you for your answer.
